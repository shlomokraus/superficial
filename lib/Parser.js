"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var parser = require("@babel/parser");
var jsondiffpatch = require("jsondiffpatch");
var prettier = require("prettier");
var Parser = /** @class */ (function () {
    function Parser() {
    }
    Parser.parse = function (src, filename, prettier) {
        if (prettier === void 0) { prettier = true; }
        if (prettier) {
            src = this.prepare(src, { filepath: filename });
        }
        return parser.parse(src, {
            allowImportExportEverywhere: true,
            sourceType: "unambiguous",
            plugins: ["classProperties", "jsx", "typescript", "objectRestSpread", "decorators-legacy"]
        });
    };
    Parser.prepare = function (src, options) {
        return prettier.format(src, options);
    };
    Parser.diff = function (src, target, ignoreKeys) {
        if (ignoreKeys === void 0) { ignoreKeys = []; }
        var diff = jsondiffpatch.create({
            propertyFilter: function (name, context) {
                return ignoreKeys.indexOf(name) < 0;
            },
            objectHash: function (obj) {
                // this function is used only to when objects are not equal by ref
                if (obj.type === "ClassProperty") {
                    console.log(obj);
                    return "ClassProperty" + obj.key.name;
                }
                return null;
            },
        });
        return diff.diff(src, target);
    };
    return Parser;
}());
exports.Parser = Parser;
