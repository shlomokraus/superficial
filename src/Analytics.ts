import UA from "universal-analytics";

export class Analytics {
  private readonly ua;

  constructor(id?: string) {
    this.ua = new UA(id);
  }

  event(...args) {
    return this.ua.event.apply(this.ua, args).send();
  }

  exception(...args) {
    return this.ua.exception.apply(this.ua, args).send();
  }

  set(...args) {
    return this.ua.set.apply(this.ua, args);
  }

  timing(...args) {
    return this.ua.timing.apply(this.ua, args).send();
  }
}
