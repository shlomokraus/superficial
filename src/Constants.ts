export const BOT_IDENTIFIER = "superficial-bot[bot]";
export const VALID_EXTENSIONS = ["ts", "js", "tsx", "jsx", "json"];
export const SCRIPT_EXTENSIONS = ["ts", "js", "tsx", "jsx"];

export enum CommentTags {
  Main = "<!-- superficial-bot-main-comment -->",
  Update = "<!-- superficial-bot-update-comment -->"
}
