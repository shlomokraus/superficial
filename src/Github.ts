import { Context } from "probot";
import { PullRequestsGetResponse } from "@octokit/rest";

export class GithubHelper
 {
  private readonly context: Context;
  private readonly pr: PullRequestsGetResponse;

  constructor(context: Context, pr: PullRequestsGetResponse) {
    this.context = context;
    this.pr = pr;
  }
  
  getFileContent = async (path: string, ref: string) => {
        const result = await this.getFile(path, ref);
        const content = Buffer.from(result.content, "base64").toString();
        return content;
      };
  

  getFile = async (path: string, ref: string) => {
    const repo = this.context.repo({ path, ref });
    const result = await this.context.github.repos.getContent(repo);
    return result.data;
  };
  
  getFiles = async () => {
    const filesRaw = await this.context.github.pullRequests.getFiles(
      this.context.repo({ number: this.pr.number})
    );
    const files = filesRaw.data.map(file => file.filename);
    return files;
  };

  getRef = async () => {
      const ref = await this.context.github.gitdata.getReference(this.context.repo({ref:"heads/"+this.pr.head.ref}));
      return ref.data;
  }

  getTree = async () => {
      const ref = await this.getRef();
      const tree = await this.context.github.gitdata.getTree(
        this.context.repo({ tree_sha: ref.object.sha })
      );
      return tree.data;
  }

  createBlob = async (content: string) => {
    const blob = await this.context.github.gitdata.createBlob(
        this.context.repo({
            number: this.pr.number,
        content,
        encoding: 'base64'
      }));
      return blob.data;
  };

  createCommit = async (files: { path: string, content: string}[]) => {

    const context = this.context;
    const github = context.github;

    const basetree = await this.getTree();
    const blobDict = {} as any;

    await Promise.all(
        files.map(async file => {
        const blob = await this.createBlob(file.content);
          blobDict[file.path] = blob.sha;
          return
    }));
    
    const tree = files.map(function(file, index) {
        return {
          path: file.path,
          mode: "100644",
          type: 'blob',
          sha: blobDict[file.path]
        };
      });

    const newTree = await github.gitdata.createTree(context.repo({
        base_tree: basetree.sha,
        tree: tree as any
    }));
      const commit = await github.gitdata.createCommit(context.repo({message: "Removing files that only have style changes", tree: newTree.data.sha, parents: [basetree.sha]}));
      await github.gitdata.updateReference(context.repo({number: this.pr.number, sha: commit.data.sha, ref: "heads/"+this.pr.head.ref }));
  };
  
}
