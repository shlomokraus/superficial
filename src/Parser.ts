const parser = require("@babel/parser");
const jsondiffpatch = require("jsondiffpatch");
const prettier = require("prettier");

export class Parser {
  static parse(src: string, filename: string, prettier = true) {
      if(prettier){
          src = this.prepare(src, { filepath: filename});
      }
    return parser.parse(src, {
      allowImportExportEverywhere: true,
      sourceType: "unambiguous",
      plugins: ["classProperties", "jsx", "typescript", "objectRestSpread", "decorators-legacy", "decorators"]
    });
  }

  static prepare(src: string, options) {
      return prettier.format(src, options);
  }

  static diff(src, target, ignoreKeys: string[] = []) {
    
    const diff = jsondiffpatch.create({
      propertyFilter: function(name, context) {
        return ignoreKeys.indexOf(name)<0;
      },
      objectHash: function(obj) {
        // this function is used only to when objects are not equal by ref
        if(obj.type==="ClassProperty"){
            console.log(obj);
            return "ClassProperty"+obj.key.name;
        }
        return null;
    },
    });
    return diff.diff(src, target);
  }

 
}
