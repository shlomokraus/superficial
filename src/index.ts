import { Application } from 'probot'

export = async (app: Application) => {
  
    app.log("Starting superficial");

}

