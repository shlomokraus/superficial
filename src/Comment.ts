import { Template, Templates } from "./Template";
import { template } from "lodash";
import { CommentTags, BOT_IDENTIFIER } from "./Constants";
import { GithubHelper } from "./Github";

// const FILE_ITEM_REGEX = /- \[(.)\] (.*)<--file-->\|/g;

export class CommentHelper {
  private readonly githubHelper: GithubHelper;

  constructor(githubHelper) {
    this.githubHelper = githubHelper;
  }

  postComment = async (
    items: { file: string; valid: boolean }[],
    errors: { file: string; error: string }[] = []
  ) => {
    const files = items.map(item => item.file);
    let comment = await this.compileComment(files, errors);

    const existing = await this.getBotMainComment();
    if (existing) {
      if (files.length === 0) {
        comment = this.tagComment(
          Template.get(Templates.updated),
          CommentTags.Main
        );
      }
      await this.githubHelper.editComment(comment, String(existing.id));
    } else {
      if (files.length > 0) {
        await this.githubHelper.createComment(comment);
      }
    }
  };

  postRevertComment = async (files: string[]) => {
    const fileItemsTemplate = Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = Template.get(Templates.revert);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    comment = this.tagComment(comment, CommentTags.Update);
    await this.githubHelper.createComment(comment);
  };

  getBotMainComment = async () => {
    const comments = await this.githubHelper.getComments();
    const filtered = comments.find(comment => {
      const isBot = comment.user.login === BOT_IDENTIFIER;
      if (!isBot) {
        return false;
      }

      const body = comment.body;
      return body.indexOf(CommentTags.Main) >= 0;
    });
    return filtered;
  };

  async compileComment(files: string[], errors: any[]) {
    const fileItemsTemplate = Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = Template.get(Templates.comment);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    const errorCount = errors.length;
    if (errorCount > 0) {
      const errTemplate = Template.get(Templates.errors);
      const errDoc = template(errTemplate)({ errors: errorCount });
      comment = comment + "\n" + errDoc;
    }

    return this.tagComment(comment, CommentTags.Main);
  }

  tagComment(body, commentTag: CommentTags) {
    if (body.indexOf(commentTag) < 0) {
      return body + "\n" + commentTag;
    }
    return body;
  }
}
