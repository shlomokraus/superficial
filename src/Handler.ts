import { Context } from "probot";
import { Parser } from "./Parser";
import fileExtension from "file-extension";
import { Template, Templates } from "./Template";
import { template } from "lodash";
import { Github } from "./Github";
const extensions = ["ts", "js", "tsx", "jsx", "json"];
const botIdentifier = "superficialbot[bot]";
// const botId = 43469792;

export class Handler {
  private readonly context: Context;
  private readonly githubHelper;
  private readonly pr;

  constructor(context: Context, pr) {
    this.context = context;
    this.githubHelper = new Github(context, pr);
    this.pr = pr;
  }

  async handle() {
    const { problematic, errors } = await this.check();

    await this.updateStatus(problematic.length === 0);
    await this.postComment(problematic, errors);

    const revertPaths = problematic.map(item => item.file);
    const revert = await Promise.all(
      revertPaths.map(async path => this.revertFile(path))
    );
    if (revert.length > 0) {
        console.log(this.githubHelper.type);
   // await this.githubHelper.createCommit(revert);
   //   await this.postRevertComment(revert.map(file => file.path));

    }
  }

  async check() {
    const files = await this.getFiles();
    const results = await Promise.all(
      files.map(async file => this.parseFile(file))
    );
    const problematic = results.filter(result => !result.valid);
    const errors = results.filter(result => result.error);
    return { problematic, errors };
  }

  parseFile = async (file: string) => {
    let valid = true;
    let error;

    try {
      const content = await this.getBaseAndHead(file);
      // If we don't have base, that is a new file so it is valid
      valid =
        content.base && content.head
          ? this.compareFiles(content.head, content.base, file)
          : true;
    } catch (ex) {
      this.context.log.error("Unable to parse file", ex);
      error = true;
    }

    return { file, valid, error };
  };

  revertFile = async (path: string) => {
    const base = this.pr.base.ref;
    const source = await this.getFile(path, base);
    return { path, content: source.content };
  };

  postRevertComment = async (files: string[]) => {
    const fileItemsTemplate = await Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = await Template.get(Templates.revert);
    const comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });
    await this.context.github.issues.createComment(
      this.context.repo({ number: this.pr.number, body: comment })
    );
  };

  postComment = async (
    items: { file: string; valid: boolean }[],
    errors: { file: string; error: string }[] = []
  ) => {
    const context = this.context;
    const files = items.map(item => item.file);
    let comment = await this.compileComment(files, errors);

    const existing = await this.getExistingComment();
    if (existing) {
      await context.github.issues.editComment(
        context.repo({
          number: this.pr.number,
          body: `_(repository was updated since posting this message)_\n ${existing.body}`,
          comment_id: String(existing.id)
        })
      );
    } else {
      if (files.length > 0) {
        await context.github.issues.createComment(
          context.repo({ number: this.pr.number, body: comment })
        );
      }
    }
  };

  getExistingComment = async () => {
    const comments = await this.context.github.issues.getComments(
      this.context.repo({number: this.pr.number})
    );
    const filtered = comments.data.find(comment => {
      const isBot = comment.user.login === botIdentifier;
      if (!isBot) {
        return false;
      }

      const body = comment.body;
      return body.indexOf("be superficial!") >= 0;
    });
    return filtered;
  };

  async compileComment(files: string[], errors: any[]) {
    const fileItemsTemplate = await Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = await Template.get(Templates.comment);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    const errorCount = errors.length;
    if (errorCount > 0) {
      const errTemplate = await Template.get(Templates.errors);
      const errDoc = template(errTemplate)({ errors: errorCount });
      comment = comment + "\n" + errDoc;
    }
    return comment;
  }

  compareFiles(source: string, target: string, filename: string) {
    const left = Parser.parse(source, filename);
    const right = Parser.parse(target, filename);
    const diff = Parser.diff(left, right, ["loc", "start", "end"]);
    return diff !== undefined;
  }

  updateStatus = async (success: boolean) => {
    const { head } = this.pr;

    const state = success ? "success" : "failure";

    function getDescription() {
      if (success) return "ready to merge";
      return "superficial changes not allowed";
    }

    let status = {
      sha: head.sha,
      state,
      description: getDescription(),
      context: "Superficial"
    };
    await this.context.github.repos.createStatus(this.context.repo(
      status
    ) as any);
  };

  getFiles = async () => {
    const filesRaw = await this.context.github.pullRequests.getFiles(
      this.context.repo({ number: this.pr.number})
    );
    const files = filesRaw.data.map(file => file.filename);
    const filtered = this.filterFiles(files);
    return filtered;
  };

  getBaseAndHead = async (path: string) => {
    let head;
    let base;
    try {
      const headRef = this.pr.head.ref;
      head = await this.getFileContent(path, headRef);
      const baseRef = this.pr.base.ref;
      base = await this.getFileContent(path, baseRef);
    } catch (ex) {
      ex = ex;
    }

    return { head, base };
  };

  getFileContent = async (path: string, ref: string) => {
    const result = await this.getFile(path, ref);
    const content = Buffer.from(result.content, "base64").toString();
    return content;
  };

  getFile = async (path: string, ref: string) => {
    const repo = this.context.repo({ path, ref });
    const result = await this.context.github.repos.getContent(repo);
    return result.data;
  };

  filterFiles(files: string[]) {
    return files.filter(file => {
      const ext = fileExtension(file);
      return extensions.indexOf(ext) >= 0;
    });
  }
}
