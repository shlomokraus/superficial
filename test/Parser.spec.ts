import { Parser } from "../src/Parser";

describe("Parser", () => {
  const src = `
    class Hello {
      a: boolean
      b: boolean
  
      render() {
          this.a = false;
          this.b = false;
          function a() {}
      }
    }
  `;

  describe("parse()", () => {
    it("Should parse source", () => {
      const parsed = Parser.parse(src, "test.ts");
      expect(parsed).toMatchSnapshot();
    });
  });

  describe("diff()", () => {
    it("Should not diff identical sources", () => {
      const parsed = Parser.parse(src, "test.ts");
      const diff = Parser.diff(parsed, parsed);
      expect(diff).toEqual(undefined);
    });

    it("Should diff non-identical sources", () => {
      const left = { a: 1 };
      const right = { a: 2 };
      const diff = Parser.diff(left, right);
      expect(diff).toEqual({ a: [1, 2] });
    });

    it("Should ignore keys", () => {
      const left = { a: 1, ignore: 1 };
      const right = { a: 2 };
      const before = Parser.diff(left, right);
      expect(before).toHaveProperty("ignore");
      const after = Parser.diff(left, right, ["ignore"]);
      expect(after).not.toHaveProperty("ignore");
    });
  });
});
