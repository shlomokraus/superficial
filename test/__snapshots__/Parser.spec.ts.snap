// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parse() Should parse source 1`] = `
Node {
  "comments": Array [],
  "end": 122,
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 11,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "body": Node {
          "body": Array [
            Node {
              "computed": false,
              "end": 27,
              "key": Node {
                "end": 17,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 3,
                    "line": 2,
                  },
                  "identifierName": "a",
                  "start": Position {
                    "column": 2,
                    "line": 2,
                  },
                },
                "name": "a",
                "start": 16,
                "type": "Identifier",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "line": 2,
                },
                "start": Position {
                  "column": 2,
                  "line": 2,
                },
              },
              "start": 16,
              "static": false,
              "type": "ClassProperty",
              "typeAnnotation": Node {
                "end": 26,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 12,
                    "line": 2,
                  },
                  "start": Position {
                    "column": 3,
                    "line": 2,
                  },
                },
                "start": 17,
                "type": "TSTypeAnnotation",
                "typeAnnotation": Node {
                  "end": 26,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 12,
                      "line": 2,
                    },
                    "start": Position {
                      "column": 5,
                      "line": 2,
                    },
                  },
                  "start": 19,
                  "type": "TSBooleanKeyword",
                },
              },
              "value": null,
            },
            Node {
              "computed": false,
              "end": 41,
              "key": Node {
                "end": 31,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 3,
                    "line": 3,
                  },
                  "identifierName": "b",
                  "start": Position {
                    "column": 2,
                    "line": 3,
                  },
                },
                "name": "b",
                "start": 30,
                "type": "Identifier",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "line": 3,
                },
                "start": Position {
                  "column": 2,
                  "line": 3,
                },
              },
              "start": 30,
              "static": false,
              "type": "ClassProperty",
              "typeAnnotation": Node {
                "end": 40,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 12,
                    "line": 3,
                  },
                  "start": Position {
                    "column": 3,
                    "line": 3,
                  },
                },
                "start": 31,
                "type": "TSTypeAnnotation",
                "typeAnnotation": Node {
                  "end": 40,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 12,
                      "line": 3,
                    },
                    "start": Position {
                      "column": 5,
                      "line": 3,
                    },
                  },
                  "start": 33,
                  "type": "TSBooleanKeyword",
                },
              },
              "value": null,
            },
            Node {
              "async": false,
              "body": Node {
                "body": Array [
                  Node {
                    "end": 75,
                    "expression": Node {
                      "end": 74,
                      "left": Node {
                        "computed": false,
                        "end": 66,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 10,
                            "line": 6,
                          },
                          "start": Position {
                            "column": 4,
                            "line": 6,
                          },
                        },
                        "object": Node {
                          "end": 64,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 8,
                              "line": 6,
                            },
                            "start": Position {
                              "column": 4,
                              "line": 6,
                            },
                          },
                          "start": 60,
                          "type": "ThisExpression",
                        },
                        "property": Node {
                          "end": 66,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 10,
                              "line": 6,
                            },
                            "identifierName": "a",
                            "start": Position {
                              "column": 9,
                              "line": 6,
                            },
                          },
                          "name": "a",
                          "start": 65,
                          "type": "Identifier",
                        },
                        "start": 60,
                        "type": "MemberExpression",
                      },
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 18,
                          "line": 6,
                        },
                        "start": Position {
                          "column": 4,
                          "line": 6,
                        },
                      },
                      "operator": "=",
                      "right": Node {
                        "end": 74,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 18,
                            "line": 6,
                          },
                          "start": Position {
                            "column": 13,
                            "line": 6,
                          },
                        },
                        "start": 69,
                        "type": "BooleanLiteral",
                        "value": false,
                      },
                      "start": 60,
                      "type": "AssignmentExpression",
                    },
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 19,
                        "line": 6,
                      },
                      "start": Position {
                        "column": 4,
                        "line": 6,
                      },
                    },
                    "start": 60,
                    "type": "ExpressionStatement",
                  },
                  Node {
                    "end": 95,
                    "expression": Node {
                      "end": 94,
                      "left": Node {
                        "computed": false,
                        "end": 86,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 10,
                            "line": 7,
                          },
                          "start": Position {
                            "column": 4,
                            "line": 7,
                          },
                        },
                        "object": Node {
                          "end": 84,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 8,
                              "line": 7,
                            },
                            "start": Position {
                              "column": 4,
                              "line": 7,
                            },
                          },
                          "start": 80,
                          "type": "ThisExpression",
                        },
                        "property": Node {
                          "end": 86,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 10,
                              "line": 7,
                            },
                            "identifierName": "b",
                            "start": Position {
                              "column": 9,
                              "line": 7,
                            },
                          },
                          "name": "b",
                          "start": 85,
                          "type": "Identifier",
                        },
                        "start": 80,
                        "type": "MemberExpression",
                      },
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 18,
                          "line": 7,
                        },
                        "start": Position {
                          "column": 4,
                          "line": 7,
                        },
                      },
                      "operator": "=",
                      "right": Node {
                        "end": 94,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 18,
                            "line": 7,
                          },
                          "start": Position {
                            "column": 13,
                            "line": 7,
                          },
                        },
                        "start": 89,
                        "type": "BooleanLiteral",
                        "value": false,
                      },
                      "start": 80,
                      "type": "AssignmentExpression",
                    },
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 19,
                        "line": 7,
                      },
                      "start": Position {
                        "column": 4,
                        "line": 7,
                      },
                    },
                    "start": 80,
                    "type": "ExpressionStatement",
                  },
                  Node {
                    "async": false,
                    "body": Node {
                      "body": Array [],
                      "directives": Array [],
                      "end": 115,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 19,
                          "line": 8,
                        },
                        "start": Position {
                          "column": 17,
                          "line": 8,
                        },
                      },
                      "start": 113,
                      "type": "BlockStatement",
                    },
                    "end": 115,
                    "generator": false,
                    "id": Node {
                      "end": 110,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 14,
                          "line": 8,
                        },
                        "identifierName": "a",
                        "start": Position {
                          "column": 13,
                          "line": 8,
                        },
                      },
                      "name": "a",
                      "start": 109,
                      "type": "Identifier",
                    },
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 19,
                        "line": 8,
                      },
                      "start": Position {
                        "column": 4,
                        "line": 8,
                      },
                    },
                    "params": Array [],
                    "start": 100,
                    "type": "FunctionDeclaration",
                  },
                ],
                "directives": Array [],
                "end": 119,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 3,
                    "line": 9,
                  },
                  "start": Position {
                    "column": 11,
                    "line": 5,
                  },
                },
                "start": 54,
                "type": "BlockStatement",
              },
              "computed": false,
              "end": 119,
              "generator": false,
              "id": null,
              "key": Node {
                "end": 51,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 8,
                    "line": 5,
                  },
                  "identifierName": "render",
                  "start": Position {
                    "column": 2,
                    "line": 5,
                  },
                },
                "name": "render",
                "start": 45,
                "type": "Identifier",
              },
              "kind": "method",
              "loc": SourceLocation {
                "end": Position {
                  "column": 3,
                  "line": 9,
                },
                "start": Position {
                  "column": 2,
                  "line": 5,
                },
              },
              "params": Array [],
              "start": 45,
              "static": false,
              "type": "ClassMethod",
            },
          ],
          "end": 121,
          "loc": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 10,
            },
            "start": Position {
              "column": 12,
              "line": 1,
            },
          },
          "start": 12,
          "type": "ClassBody",
        },
        "end": 121,
        "id": Node {
          "end": 11,
          "loc": SourceLocation {
            "end": Position {
              "column": 11,
              "line": 1,
            },
            "identifierName": "Hello",
            "start": Position {
              "column": 6,
              "line": 1,
            },
          },
          "name": "Hello",
          "start": 6,
          "type": "Identifier",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 10,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "superClass": null,
        "type": "ClassDeclaration",
      },
    ],
    "directives": Array [],
    "end": 122,
    "interpreter": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 0,
        "line": 11,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "script",
    "start": 0,
    "type": "Program",
  },
  "start": 0,
  "type": "File",
}
`;
